File: submissions/0278-first-bad-version/solution.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        //System.out.println("---------:" + n);
        return binaryFind(1, n);
    }
    
    private int binaryFind(int left, int right) {
        if(left>right) {
            //System.out.println("----left:" + left +", right:" + right);
            return left + (isBadVersion(left) ? 0 : 1);
        }
        int current = left + (right-left)/2;
        //System.out.println("----left:" + left +", right:" + right + ", current:" + current);
        if(isBadVersion(current)) {
            return binaryFind(left, current-1);
        } else {
            return binaryFind(current+1, right);
        }
    }
}

File: submissions/0278-first-bad-version/solution.py
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

class Solution:
    def firstBadVersion(self, n: int) -> int:
        result = -1

        lo = 1
        hi = n+1
        target = -1

        while lo < hi:
            mid = (lo + hi) // 2
            if not isBadVersion(mid):
                lo = mid + 1
            else:
                target = mid
                hi = mid

        return target

        
